File: Section-2/CodeSrc/StringType/ReverseWordsinaStringIII.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Utils.Log;

namespace CodeSrc.StringType
{
    class ReverseWordsinaStringIII
    {
        private const string tag = "ReverseWordsinaStringIII";

        public string ReverseWords(string s)
        {
            if (s == null || s.Equals(" "))
            {
                return s;
            }

            char[] chArr = s.ToArray();
            StringBuilder sb = new StringBuilder();
            Stack<char> stack = new Stack<char>();

            foreach (char ch in chArr)
            {
                if (ch.Equals(' '))
                {
                    while (stack.Count != 0)
                    {
                        sb.Append(stack.Pop());
                    }
                    sb.Append(ch);
                }
                else
                {
                    stack.Push(ch);
                }
            }

            while (stack.Count != 0)
            {
                sb.Append(stack.Pop());
            }

            return sb.ToString();
        }

        public static void Main(string[] args)
        {
            string msg = "Let's take LeetCode contest";
            LogUtils.D(tag, new ReverseWordsinaStringIII().ReverseWords(msg));
        }
    }
}

File: src/main/java/lizbyu/stack/ReverseWordsinaStringIII.java
package lizbyu.stack;

import java.util.Deque;
import java.util.LinkedList;

public class ReverseWordsinaStringIII {
    public String reverseWords(String s) {
        Deque<Character> stack = new LinkedList<>();
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < s.length(); i++) {
            char target = s.charAt(i);
            if (target == ' ') {
                while (!stack.isEmpty()) {
                    sb.append(stack.pop());
                }
                sb.append(' ');
            } else {
                stack.push(target);
            }
            if (i == s.length() - 1) {
                while (!stack.isEmpty()) {
                    sb.append(stack.pop());
                }
            }
        }

        return sb.toString();
    }
}
